### PR TITLE
Modernize the FIR module to support optimized smooth scaling.

### PR DESCRIPTION
**Pull request by kwagyeman**
* Adds hmirror, vflip, and transpose to read_ir.
* Improves memory allocation of read_ir such that it's not trashing the heap anymore.
* Peak memory usage reduced.
* Refactored code to improve things.
* Made snapshot() support all the same args as read_ir/draw_ir with the ability to set custom output resolutions.

This PR significantly improves the performance of the FIR module by using tight'ish loops for all operations and it avoids copying IR arrays into the heap to MP land and then back into C land again.

All sensors are tested and working.

### CODE REVIEW

File: src/omv/modules/py_fir.c [hunk 1]
```diff
@@ -78,352 +145,385 @@ static void fir_fill_image_float_obj(image_t *img, mp_obj_t *data, float min, fl
...
-                xfree(mlx_data);
-                mlx_data = NULL;
-                goto FIR_MLX90621;
+            if (error != 0) {
+                if (first_init) {
+                    first_init = false;
+                    cambus_pulse_scl(&fir_bus);
+                    goto FIR_MLX90621_RETRY;
+                } else {
+                    nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Failed to init the MLX90621!"));
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:314]:
```suggestion
                } else {
                    cambus_deinit(&fir_bus);
                    nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Failed to init the MLX90621!"));
```
Deinit the bus here ?

File: src/omv/modules/py_fir.c [hunk 2]
```diff
@@ -607,62 +794,83 @@ mp_obj_t py_fir_snapshot(uint n_args, const mp_obj_t *args, mp_map_t *kw_args)
...
-        arg_other->bpp = image.bpp;
+    switch(fir_sensor) {
+        case FIR_MLX90621: {
+            float Ta, *To = fb_alloc(MLX90621_WIDTH * MLX90621_HEIGHT * sizeof(float), FB_ALLOC_NO_HINT);
+            fir_MLX90621_get_frame(&Ta, To);
+            if (!scale_obj) fir_get_minmax(To, MLX90621_WIDTH * MLX90621_HEIGHT, &min, &max);
+            fir_fill_image_float(&src_img, MLX90621_WIDTH, MLX90621_HEIGHT, To, min, max,
+                    arg_hmirror ^ true, arg_vflip, arg_transpose, true);
+            fb_free();
+            break;
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:828]:
```suggestion
            float Ta;
            float *To = fb_alloc(MLX90621_WIDTH * MLX90621_HEIGHT * sizeof(float), FB_ALLOC_NO_HINT);
            fir_MLX90621_get_frame(&Ta, To);
            if (!scale_obj) {
                fir_get_minmax(To, MLX90621_WIDTH * MLX90621_HEIGHT, &min, &max);
            }
            fir_fill_image_float(&src_img, MLX90621_WIDTH, MLX90621_HEIGHT, To, min, max,
                    arg_hmirror ^ true, arg_vflip, arg_transpose, true);
            fb_free();
            break;
```
Formatting.

File: src/omv/modules/py_fir.c [hunk 3]
```diff
@@ -607,62 +794,83 @@ mp_obj_t py_fir_snapshot(uint n_args, const mp_obj_t *args, mp_map_t *kw_args)
...
 STATIC const mp_rom_map_elem_t globals_dict_table[] = {
     { MP_ROM_QSTR(MP_QSTR___name__),        MP_OBJ_NEW_QSTR(MP_QSTR_fir) },
     { MP_ROM_QSTR(MP_QSTR_FIR_NONE),        MP_ROM_INT(FIR_NONE) },
-    { MP_ROM_QSTR(MP_QSTR_FIR_SHIELD),      MP_ROM_INT(MLX90621) },
-    { MP_ROM_QSTR(MP_QSTR_FIR_MLX90620),    MP_ROM_INT(MLX90621) },
-    { MP_ROM_QSTR(MP_QSTR_FIR_MLX90621),    MP_ROM_INT(MLX90621) },
+    { MP_ROM_QSTR(MP_QSTR_FIR_SHIELD),      MP_ROM_INT(FIR_MLX90621) },
+    { MP_ROM_QSTR(MP_QSTR_FIR_MLX90621),    MP_ROM_INT(FIR_MLX90621) },
     { MP_ROM_QSTR(MP_QSTR_FIR_MLX90640),    MP_ROM_INT(FIR_MLX90640) },
     { MP_ROM_QSTR(MP_QSTR_FIR_AMG8833),     MP_ROM_INT(FIR_AMG8833) },
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:868]:
```suggestion
    { MP_ROM_QSTR(MP_QSTR_FIR_SHIELD),      MP_ROM_INT(FIR_MLX90621) },
    { MP_ROM_QSTR(MP_QSTR_FIR_MLX90621),    MP_ROM_INT(FIR_MLX90621) },
    { MP_ROM_QSTR(MP_QSTR_FIR_MLX90640),    MP_ROM_INT(FIR_MLX90640) },
    { MP_ROM_QSTR(MP_QSTR_FIR_AMG8833),     MP_ROM_INT(FIR_AMG8833) },
        // Color Palettes
    { MP_OBJ_NEW_QSTR(MP_QSTR_PALETTE_RAINBOW), MP_OBJ_NEW_SMALL_INT(COLOR_PALETTE_RAINBOW)},
    { MP_OBJ_NEW_QSTR(MP_QSTR_PALETTE_IRONBOW), MP_OBJ_NEW_SMALL_INT(COLOR_PALETTE_IRONBOW)},
```

**Review by iabdalkader — CHANGES_REQUESTED** [review]

File: scripts/examples/OpenMV/12-Thermopile-Shield/AMG8833_camera.py [hunk 4]
```diff
@@ -15,7 +19,9 @@
     clock.tick()
 
     try:
-        img = fir.snapshot(copy_to_fb=True)
+        img = fir.snapshot(x_size=w, y_size=h,
+                           color_palette=sensor.PALETTE_IRONBOW, hint=drawing_hint,
+                           copy_to_fb=True)
```
[the remark this answers is not in the document]
Comment (kwagyeman) [reply scripts/examples/OpenMV/12-Thermopile-Shield/AMG8833_camera.py:24]: Done

File: scripts/examples/OpenMV/12-Thermopile-Shield/AMG8833_camera.py [hunk 5]
```diff
@@ -3,10 +3,14 @@
...
 
-import image, time, fir
+import sensor, image, time, fir
+
+drawing_hint = image.BICUBIC # or image.BILINEAR or 0 (nearest neighbor)
 
 # Initialize the thermal sensor
 fir.init(type=fir.FIR_AMG8833)
+w = fir.width() * 20
+h = fir.height() * 20
```
[the remark this answers is not in the document]
Comment (kwagyeman) [reply scripts/examples/OpenMV/12-Thermopile-Shield/AMG8833_camera.py:13]: I don't want to do this right now. Here are my reasons:

1. The actual resolution of the sensors are like 8x8 or 16x4 or 32x24. Nothing near any of the standard frame sizes.
2. Bicubic scaling is just for making pretty images. One should not use it for computer vision processing. As such, the user should be aware that they are scaling the image up using bicubic scaling.
3. Our current frame sizes don't really match with the sensor sizes.

That said, I am not against adding this in the future. But, not in this PR.

File: src/omv/modules/py_fir.c [hunk 6]
```diff
@@ -78,352 +145,385 @@ static void fir_fill_image_float_obj(image_t *img, mp_obj_t *data, float min, fl
...
+static inline int16_t fir_AMG8833_12_to_16(int16_t value)
 {
-    width = 0;
-    height = 0;
-    ADC_resolution = 0;
-    IR_refresh_rate = 0;
-    if (mlx_data != NULL) {
-        mlx_data = NULL;
+    if ((value >> 11) & 1) value |= 1 << 15;
+    return value & 0x87FF;
```
[the remark this answers is not in the document]
Comment (kwagyeman) [reply src/omv/modules/py_fir.c:182]: ... The above is not the same as what I wrote.

File: src/omv/modules/py_fir.c [hunk 7]
```diff
@@ -78,352 +145,385 @@ static void fir_fill_image_float_obj(image_t *img, mp_obj_t *data, float min, fl
...
-            int error = 0;
-            error |= MLX90621_DumpEE(eeprom);
+            uint8_t *eeprom = fb_alloc(MLX90621_EEPROM_DATA_SIZE * sizeof(uint8_t), FB_ALLOC_NO_HINT);
+            int error = MLX90621_DumpEE(eeprom);
             error |= MLX90621_Configure(eeprom);
-            error |= MLX90621_SetResolution(ADC_resolution);
-            error |= MLX90621_SetRefreshRate(IR_refresh_rate);
-            error |= MLX90621_ExtractParameters(eeprom, mlx_data);
+            error |= MLX90621_SetRefreshRate(ir_fresh_rate);
+            error |= MLX90621_SetResolution(adc_resolution);
```
[the remark this answers is not in the document]
Comment (kwagyeman) [reply src/omv/modules/py_fir.c:304]: I tested it.

File: src/omv/modules/py_fir.c (hunk 1 above)
Comment (kwagyeman) [reply src/omv/modules/py_fir.c:314]: The exception causes init0 to be run which deinits the bus. However, should just call py_fir_deinit.

File: src/omv/modules/py_fir.c [hunk 8]
```diff
@@ -8,50 +8,117 @@
...
+
+    for (int i = 0; i < len; i++) {
+        float temp = data[i];
+        if (temp < min) min = temp;
+        if (temp > max) max = temp;
+    }
+
+    *p_min = min;
+    *p_max = max;
+}
```
[the remark this answers is not in the document]
Comment (kwagyeman) [reply src/omv/modules/py_fir.c:74]: Done

File: src/omv/modules/py_fir.c [hunk 9]
```diff
@@ -8,50 +8,117 @@
...
+                float raw = raw_row[y];
+                if (raw < min) raw = min;
+                if (raw > max) raw = max;
+                int pixel = fast_roundf((raw - min) * diff);
+                pixel = __USAT(pixel, 8);
+                if (!dst_transpose) row_pointer[y_dst * w] = pixel;
+                else t_row_pointer[y_dst] = pixel;
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (kwagyeman) [reply src/omv/modules/py_fir.c]: Fixed

File: src/omv/modules/py_fir.c (hunk 1 above)
Comment (kwagyeman) [reply src/omv/modules/py_fir.c:288]: Added the deinit function.

File: src/omv/modules/py_fir.c [hunk 10]
```diff
@@ -583,15 +661,102 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_KW(py_fir_draw_ir_obj, 2, py_fir_draw_ir);
...
+        if (kw_arg && MP_OBJ_IS_TYPE(kw_arg->value, mp_const_none)) {
+            color_palette = NULL;
+        } else if ((n_args > arg_index) && MP_OBJ_IS_TYPE(args[arg_index], mp_const_none)) {
+            color_palette = NULL;
+        } else if (py_helper_keyword_int_maybe(n_args, args, arg_index, kw_args, MP_OBJ_NEW_QSTR(MP_QSTR_color_palette), &palette)) {
+            if (palette == COLOR_PALETTE_RAINBOW) color_palette = rainbow_table;
+            else if (palette == COLOR_PALETTE_IRONBOW) color_palette = ironbow_table;
+            else nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Invalid pre-defined color palette!"));
+        } else {
+            image_t *arg_color_palette = py_helper_keyword_to_image_mutable_color_palette(n_args, args, arg_index, kw_args);
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:708]: Please fix the code formatting.

File: src/omv/modules/py_fir.c [hunk 11]
```diff
@@ -583,15 +661,102 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_KW(py_fir_draw_ir_obj, 2, py_fir_draw_ir);
...
+
+    const uint8_t *alpha_palette = NULL;
+    {
+        image_t *arg_alpha_palette = py_helper_keyword_to_image_mutable_alpha_palette(n_args, args, 9, kw_args);
+
+        if (arg_alpha_palette) {
+            if (arg_alpha_palette->bpp != IMAGE_BPP_GRAYSCALE) nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Alpha palette must be GRAYSCALE!"));
+            if ((arg_alpha_palette->w * arg_alpha_palette->h) != 256) nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Alpha palette must be 256 pixels!"));
+            alpha_palette = (uint8_t *) arg_alpha_palette->data;
+        }
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:726]: Please fix the code formatting.

File: src/omv/modules/py_fir.c [hunk 12]
```diff
@@ -583,15 +661,102 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_KW(py_fir_draw_ir_obj, 2, py_fir_draw_ir);
...
+    if (got_x_scale && got_x_size) nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Choose either x_scale or x_size not both!"));
+    if (got_y_scale && got_y_size) nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Choose either y_scale or y_size not both!"));
+
+    if (got_x_size) arg_x_scale = arg_x_size / ((float) arg_roi.w);
+    if (got_y_size) arg_y_scale = arg_y_size / ((float) arg_roi.h);
+
+    if ((!got_x_scale) && (!got_x_size) && got_y_size) arg_x_scale = arg_y_scale;
+    if ((!got_y_scale) && (!got_y_size) && got_x_size) arg_y_scale = arg_x_scale;
+
+    mp_obj_t scale_obj = py_helper_keyword_object(n_args, args, 13, kw_args, MP_OBJ_NEW_QSTR(MP_QSTR_scale), NULL);
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:746]: Please fix the code formatting.

File: src/omv/modules/py_fir.c [hunk 13]
```diff
@@ -583,15 +661,102 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_KW(py_fir_draw_ir_obj, 2, py_fir_draw_ir);
...
+
+    if (scale_obj) {
+        mp_obj_t *arg_scale;
+        mp_obj_get_array_fixed_n(scale_obj, 2, &arg_scale);
+        min = mp_obj_get_float(arg_scale[0]);
+        max = mp_obj_get_float(arg_scale[1]);
+    }
+
+    int arg_pixformat = py_helper_keyword_int(n_args, args, 14, kw_args, MP_OBJ_NEW_QSTR(MP_QSTR_pixformat), PIXFORMAT_RGB565);
+    if ((arg_pixformat != PIXFORMAT_GRAYSCALE) && (arg_pixformat != PIXFORMAT_RGB565)) nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Invalid pixformat!"));
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:757]: fix formatting

File: src/omv/modules/py_fir.c [hunk 14]
```diff
@@ -583,15 +661,102 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_KW(py_fir_draw_ir_obj, 2, py_fir_draw_ir);
...
+    bool got_x_scale = py_helper_keyword_float_maybe(n_args, args, 3, kw_args, MP_OBJ_NEW_QSTR(MP_QSTR_x_scale), &arg_x_scale);
+
+    float arg_y_scale = 1.f;
+    bool got_y_scale = py_helper_keyword_float_maybe(n_args, args, 4, kw_args, MP_OBJ_NEW_QSTR(MP_QSTR_y_scale), &arg_y_scale);
+
+    rectangle_t arg_roi;
+    py_helper_keyword_rectangle_roi(&src_img, n_args, args, 5, kw_args, &arg_roi);
+
+    int arg_rgb_channel = py_helper_keyword_int(n_args, args, 6, kw_args, MP_OBJ_NEW_QSTR(MP_QSTR_rgb_channel), -1);
+    if ((arg_rgb_channel < -1) || (2 < arg_rgb_channel)) nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "-1 <= rgb_channel <= 2!"));
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:688]: formatting

File: src/omv/modules/py_fir.c [hunk 15]
```diff
@@ -583,15 +661,102 @@ STATIC MP_DEFINE_CONST_FUN_OBJ_KW(py_fir_draw_ir_obj, 2, py_fir_draw_ir);
 mp_obj_t py_fir_snapshot(uint n_args, const mp_obj_t *args, mp_map_t *kw_args)
 {
     if (fir_sensor == FIR_NONE) return mp_const_none;
```
Comment (iabdalkader) [inline src/omv/modules/py_fir.c:663]: Formatting

**Review by iabdalkader — CHANGES_REQUESTED** [review]

**Pull request merged**